pkg/tfbridge/x/tokens.go

```diff
@@ -649,62 +649,96 @@ func aliasResource(
 	if res == nil {
 		return
 	}
-
-	safeRange := func(r shim.Resource, f func(string, shim.Schema)) {
+	var walk func(*fieldHistory, *b.SchemaInfo, shim.Schema) (bool, bool)
+	walkResource := func(r shim.Resource, hist *map[string]*fieldHistory, info *map[string]*b.SchemaInfo) (bool, bool) {
 		if r == nil {
-			return
+			return hist != nil, info != nil
 		}
 		m := r.Schema()
 		if m == nil {
-			return
+			return hist != nil, info != nil
 		}
+
+		var rHasH, rHasI bool
+
 		m.Range(func(k string, v shim.Schema) bool {
-			f(k, v)
+			h, hasH := getNonNil(hist, k)
+			i, hasI := getNonNil(info, k)
+			keepH, keepI := walk(h, i, v)
```

Comment:
I'm not sure what is being kept here.

_Should_ we keep history/info?
Are we checking _whether_ history or info are being kept?

Reply:
We should keep it when (1) it already existed or (2) we stored useful information in it. I added some comments explaining what we are trying to do that will hopefully clarify.